Rewrite ConnectionReuse

gnunicorn opened the pull request:
based upon #355 but has a centralised `ConnectionManager`, which encapsulates  and manages all state transitions in a few central places and ensures everyone is notified as needed if those occur. Which means:

 - [x] No one outside of it knows about the `PeerState`-Enum whatsoever
 - [x] Everyone just references the `Arc` not the `Mutex`, the Managers manages all locking
 - [x] Independent locks for `connections` and `counters`
 - [x] More specific locks around the `connections` with less code and more explicit transitions, making it easier to reason about the code 
 - [x] Ensures `Counters` are increased properly (rather than relying on external parties to do that)
 - [x] No `Poisoned`-State possible anymore (and removed alltogether)
 - [x] Remove `connection_id`-system and consequently all referencing of `ConnectionManager` in Streams, replace with shared `closed`-state to stop streams early
 - [x] Add `reset_connection` on `ConnectionReuseSubstream` to allow a substream to state that a connection is to be considered closed (primarily for the `Ping`-protocol) and doesn't continue producing
 - [x] which all allowed removing many unnecessary fields of `PeerState`

Further more:
  - [x] remove substream-referencing of `ConnectionManager` alltogether, massivley reduxoing
  - [x] [Performance] Replaced `FnvHashMap<usize, task::Task>`, with a faster `Vec<(usize, task::Task)>`
  - [x] [Fix] When an incoming connection replaced a pending state, we didn't inform all listeners, we now close all streams
 - [x] [Fix] reliably ensure we aren't keeping around/use old potentially stale connections, fixes a leak and potential DoS-Attack-Window.

gnunicorn commented on core/src/connection_reuse.rs:
```diff
@@ -40,193 +40,541 @@
-    // Struct shared between most of the `ConnectionReuse` infrastructure.
-    shared: Arc<Mutex<Shared<C::Output>>>,
+    /// Tasks to notify when one or more new elements were added to `connections`.
+    notify_on_new_connec: Mutex<Vec<(usize, task::Task)>>,
```
We were wondering about the performance of this, so [I made a bunch of test](https://gist.github.com/gnunicorn/24d513debae25bfd3586e01e58f7dbd0). They showed that the overhead of hashing is only cut of once the `vec` is actually quite large (>200) - which is super unlikely to happen for the notifications (as barely ever more than a handfull of tasks would want to be notified, and we drain it every time we do). What makes the HashMap worse in this scenario, is that we "intend-to-insert" much more often than we actually insert, and every time we do, we do the expensive hashing in the hashmap while we'd exit early in looping through the vector. In practice the vector has one or two entries, making it faster to loop than to hash (even with `Fnv`). I was told these supposedly tiny performance differences matter, so I made the change.

twittner replied:
I do not think these benchmarks do the same thing.  For instance [`bench_custom_hasher_hashmap_same`](https://gist.github.com/gnunicorn/24d513debae25bfd3586e01e58f7dbd0#file-fnv_bench-rs-L77) does `COUNT` inserts, whereas [`bench_with_vec_same`](https://gist.github.com/gnunicorn/24d513debae25bfd3586e01e58f7dbd0#file-fnv_bench-rs-L165) returns early, i.e. after `MODULO` inserts. Changing `bench_with_vec_same` to do (roughly) what the `HashMap` based benchmark does


```rust
#[bench]
fn bench_with_vec_same(b: &mut Bencher) {
    b.iter(|| {
        let n = test::black_box(COUNT);
        let mut v: Vec<(Key, &'static str)> = vec![];
        'outer: for idx in 0..n {
            let id = idx % MODULO;
            for i in 0..v.len() {
                if v.get(i).unwrap().0 == id {
                    v[i] = (id, "value");
                    continue 'outer;
                }
            }
            v.push((id, "value"));
        };
    });
}
```

yields

```
running 2 tests
test bench_custom_hasher_hashmap ... bench:       8,800 ns/iter (+/- 376)
test bench_with_vec_same         ... bench:      32,970 ns/iter (+/- 1,188)
```

gnunicorn replied:
@twittner oh, yes, what a stupid mistake - I had an iterator in there first and never replaced that `return`. I have to run these tests again.

why are you doing a `v[i] = (id, "value");` if you found it though?

gnunicorn replied:
Sorry about that, that was a stupid bug. However, I inserted the `Modulo` only to emulate a scenario closer to our "many attempts little inserts" scenario. I only tried that because the worst-case-scenario (`never matches, inserts every time`) yielded good results for smaller sizes, which was unaffected by the bug. I fixed the code and running it now on [small sets still yields significant performance improvemnets](https://gist.github.com/gnunicorn/24d513debae25bfd3586e01e58f7dbd0#file-_results-txt-L2). On my system - take it with a grain of salt, flexible cpu-rate and all - the `HashMap` only shows benefit after 30+ items, which we'll probably not reach in practice:

```
# 500 Modulo 30
test bench_custom_hasher_hashmap_same ... bench:       5,626 ns/iter (+/- 608)
test bench_fnv_hashmap_same           ... bench:       8,043 ns/iter (+/- 3,692)
test bench_fnv_hashset_same           ... bench:       6,652 ns/iter (+/- 394)
test bench_fnv_hashset_uniques        ... bench:      12,490 ns/iter (+/- 2,910)
test bench_with_sorted_vec_same       ... bench:      11,130 ns/iter (+/- 457)
test bench_with_vec_same              ... bench:       6,454 ns/iter (+/- 154)

test result: ok. 0 passed; 0 failed; 0 ignored; 6 measured; 0 filtered out

```

twittner replied:
I can not reproduce those results. `bench_custom_hasher_hashmap_same` is always faster or on par with `bench_with_vec_same` on my machine, so I would prefer to go with Big-O and use a `HashMap`.

```
# 100 Modulo 20
running 2 tests
test bench_custom_hasher_hashmap_same ... bench:         909 ns/iter (+/- 62)
test bench_with_vec_same              ... bench:       1,026 ns/iter (+/- 59)

# 200 Modulo 20
running 2 tests
test bench_custom_hasher_hashmap_same ... bench:       1,782 ns/iter (+/- 74)
test bench_with_vec_same              ... bench:       1,903 ns/iter (+/- 93)

# 2000 Modulo 20
running 2 tests
test bench_custom_hasher_hashmap_same ... bench:      17,063 ns/iter (+/- 366)
test bench_with_vec_same              ... bench:      17,590 ns/iter (+/- 933)

# 1000 Modulo 100
running 2 tests
test bench_custom_hasher_hashmap_same ... bench:       9,126 ns/iter (+/- 448)
test bench_with_vec_same              ... bench:      38,695 ns/iter (+/- 1,522)

# 500 Modulo 30
running 2 tests
test bench_custom_hasher_hashmap_same ... bench:       4,605 ns/iter (+/- 237)
test bench_with_vec_same              ... bench:       7,246 ns/iter (+/- 471)
```

gnunicorn replied:
@twittner and then use the custom hasher? Fine with me. I am not sure my results aren't influenced because of the automatic CPU-frequency-govener that I can't disable :(  . Especially for the first tests this probably skewes the result as it will only ramp up CPU freq during the test... That the custom hasher (which does nothing) is supposedly magnitutes slower than the Fnv supports that thesis - because Fnv doesn't do anything else than we do...

I'll do that.

tomaka replied:
I think `SmallVec` should be faster because of cache locality.

gnunicorn commented:
@twittner yes, indeed, _it intended to_. However, it has/had a bug and a major drawback:
1. when a new connection came in, it would just overwrite any potentially existing `PeerState`-Entry with a new one - if that was an active one, it would leaving all existing connections dangling. Given, it is a bit weird they connect if we are already connected, but then, they might do it because the connection timed and we  just don't know yet. While this also contains a potential leak attack vector, in practice most streams would time out and then be dropped, decreasing the counter - a counter, we just recently reseted, and it didn't actually increase - thus leading to the connection being dropped much earlier than it should be as the counter reaches zero in the next drop. Annoying.
2. through that system no stream could signal the connection to forcefully reset, all you could do is drop this stream and hope that others drop their streams, too. The use case here is that the Ping-Protocol should have a way to signal that the entire connection must be dropped and no other should continue to read or write, once we detected it to be down. With the simple counter this was impossible to realise unless all streams would check that counter every time they are asked to perform an action, leading to quite a performance penalty. (In the same vain, the counter lead to a much more complex `PeerState` and `Substream`-Struct)

twittner commented:
> no stream could signal the connection to forcefully reset

Removing a connection from the cache does not necessarily close or reset the connection. Depending on the `StreamMuxer` implementation, each `Substream` [may have a reference to the underlying connection.](https://github.com/libp2p/rust-libp2p/blob/5546bb9dfdd57b308479c266c944318574e7290e/mplex/src/lib.rs#L297) To properly implement connection resets the `StreamMuxer` trait would need to have a `close` or `reset` method added.

gnunicorn commented:
@twittner 

> To properly implement connection resets the StreamMuxer trait would need to have a close or reset method added.

Agreed. However, with `type Substream: AsyncRead + AsyncWrite;` being the only thing we return to many and neither of them having `close` or `reset` functions, I added the shared `closed: Arc<AtomicBool>`, which anyone with access to `ConnectionReuseSubstream` can set through calling `reset_connection` on it. It will still not properly clean up the connections but rather renders them useless, so we assume they are being dropped because of that. But you are correct, this isn't the cleanest way and we might want to consider an official "closing"-functionality.

gnunicorn commented:
@twittner addressed your concerns.

tomaka commented on core/src/connection_reuse.rs:
```diff
@@ -40,193 +40,543 @@
+use std::hash::{BuildHasherDefault, Hasher};
+
+#[derive(Default)]
+struct UsizeHasher {
```
Why not use `FnvHashMap`, which afaik does this better, instead of defining our own hasher?

gnunicorn replied:
Performance. [See the conversation I had with @twittner about this](https://github.com/libp2p/rust-libp2p/pull/387#discussion_r208851171). I am not sure what you mean with `FnvHashMap` doing "this better", `FnvHasher` does the same we do here: [it defines its own hasher](https://doc.servo.org/src/fnv/lib.rs.html#114-121), using the Fnv-Algo. However as we already have pretty unique values, that Algo is a bit more expensive (~10-15% in this case) than us just converting the `usize`-Key into the required `u64` type. We are doing the exact same as `Fnv` but with a less expensive "algorithm" (if you can call it that). 

This algorithm however is very much constraint to the use case at hand and should probably not be re-used at many other places. Which is why I can understand the argument to go for a more generic solution - if we are willing to pay the performance price, I am fine with doing that, too.

tomaka commented on core/src/connection_reuse.rs:
```diff
@@ -40,193 +40,543 @@
+/// Knows about all connections and their current state, allows one to poll for
+/// incoming and outbound connections, while managing all state-transitions that
+/// might occur automatically.
+// #[derive(Clone)]
```
Let's remove this comment. This struct doesnt' need to be cloned.

tomaka commented on core/src/connection_reuse.rs:
```diff
@@ -40,193 +40,543 @@
+    /// substreams. Consume the PeerState while doing so.
+    fn reset_conn(self) {
+        if let PeerState::Active { closed, .. } = self {
+            closed.store(false, Ordering::Release);
```
I'm not a fan of having `Ordering::Release` without any context, as the ordering to use strongly depends on the context.
Personally I'd just copy paste these three lines of code at the two places where we use them.

gnunicorn replied:
I generally agree with you regarding copying three lines around. The reason I went for this being a constraint within a function rather than copied is that this isn't yet fully resetting the connection yet. So, I can see us quickly expanding what should happen here (by first adding some debugging statements, later doing some actual signaling/cleaning up of the related connections) and if you then had to do that at many code location, you easily miss one when doing changes and most reviewer(s) won't notice. If you insist, though, I don't have a strong opinion on this and can do a copy instead.

Regarding the `Ordering:Release`, I am not sure what else we would want to do here/at the specific locations. @twittner pointed out correctly, [that `Ordering:Relaxed` might simply never gets propagated](https://github.com/libp2p/rust-libp2p/pull/387#discussion_r208512061). I don't think that we need to expect a strong guarantee of ordering in this case, as it doesn't actually hurt if you read the remaining bytes of an already closed connection. The only important things here, to me is, that it will _eventually_ fail and not dangle around for too long.

tomaka commented on core/src/connection_reuse.rs:
```diff
@@ -40,193 +40,543 @@
-    active_connections: FnvHashMap<Multiaddr, M>,
+    /// Notify all `Tasks` in `self.notify_on_new_connec` that a new connection was established
+    /// consume the tasks in that process
+    fn new_con_notify(&self) {
```
I commented about this in private, but I'm really not a fan of locking a mutex in a function that is intended to be called from another function which can potentially lock a mutex as well.
This greatly increases the difficulty of the reviewer if you want to make sure that there is no deadlock.

gnunicorn replied:
> I commented about this in private, but I'm really not a fan of locking a mutex in a function that is intended to be called from another function which can potentially lock a mutex as well.

While I agree on the general merit of this statement, I don't think this applies here: There are two distinct `Mutex`s, each one is clearly bound to corresponding function calls. I don't really see how it gives us any benefit if instead, we'd `Mutex` the entire `ConnectionManager` (what we had before) and therefore prevent anyone using the `new_notify_registration` (the only function actually used from the outside) until we are done with the respective `poll_`s. Even in that scenario, if `task.1.notify` _would_ cause an immediate `poll` (which is the only side-effect function and we know _it doesn't_ as this wake-up is only queued) we'd deadlock. 

I totally agree with you that `Mutex`s make the reviewer harder, as you need to make sure there isn't any side-effect code within the `lock`ed code that might cause a `deadlock`. Which is exactly why I try to keep the code between locks as constraint, short and side-effect-free as possible.

I am fine with moving this lock+drain on the respective position in the code base (as it is only used once) but I see little to no merit in putting the entire object into a mutex just for these two clearly independent locks... If you insist however, I can also do that.

tomaka commented on core/src/connection_reuse.rs:
```diff
@@ -40,193 +40,543 @@
+    fn clear_error(&self, addr: Multiaddr) -> bool {
+        let mut conns = self.connections.lock();
+
+        if let Entry::Occupied(e) = conns.entry(addr) {
```
Should be `get()` instead of `entry()` so that `clear_error` can take `addr` by reference.

tomaka commented on core/src/connection_reuse.rs:
```diff
@@ -40,193 +40,543 @@
+    }
+}
+
+fn poll_for_substream<M>(mut outbound: M::OutboundSubstream, addr: &Multiaddr)
```
What does this function do?
I don't care if a function doesn't have documentation if it is trivial, but this function clearly is not.

gnunicorn commented:
@tomaka addressed your remarks. Simplified code, fixed or commented on.

I am not fond of this (or anything) to be only understood by one person (no matter how many tests we have). I totally agree this is still super complicated (code-wise) and I am happy to discuss how we can simplify it further. Do you have any ideas?

tomaka commented:
> Do you have any ideas?

Extract code into submodules.
However the splitting point is not easy to find.

gnunicorn commented:
Decided to try a totally different approach. Pausing.

tomaka commented:
Could you fix the conflicts?

tomaka commented:
I updated with the changes in master.

tomaka commented on core/src/connection_reuse.rs:
```diff
@@ -40,205 +40,556 @@
+                        }
+                        let muxer = Arc::new(muxer);
+                        match poll_for_substream(muxing::outbound_from_ref_and_wrap(muxer.clone()), &addr) {
+                            Ok(None) => return Ok(Async::NotReady),
```
If we successfully dial a remote but it's not ready to open a substream yet, this means we will poll the dial again the next time and probably panic.

tomaka commented on core/src/connection_reuse.rs:
```diff
@@ -40,205 +40,556 @@
+
+/// Polls the given outbound of addr for a new substream. Returns `Ok(Some(stream))` if available,
+/// `Ok(None)` if the stream is not ready yet add an `Err` if the stream had failed.
+fn poll_for_substream<O, S>(mut outbound: O, addr: &Multiaddr)
```
Why isn't that a `Poll`, so that we return `Async::Ready` and `NotReady` instead of respectively `Some` and `None`?

tomaka commented on core/src/connection_reuse.rs:
```diff
@@ -40,205 +40,556 @@
+    <UpgradedNode<T, C> as Transport>::MultiaddrFuture: 'static,
+{
+    /// Informs the Manager about a new inbound connection that has been
+    /// established, so it
```
so it?

tomaka commented on core/src/connection_reuse.rs:
```diff
@@ -40,205 +40,556 @@
+                task.notify()
+            }
+        } else {
+            old_state.map(|s| s.reset_conn());
```
This is very bad style in my opinion. You're mixing a "pulling" API (`map`) with a "pushing" action (`reset_conn`).

tomaka commented on core/src/connection_reuse.rs:
```diff
@@ -40,205 +40,556 @@
+                        "Using existing connection to {} to open outbound substream",
+                        addr
+                    );
+                    match poll_for_substream(muxing::outbound_from_ref_and_wrap(muxer.clone()), &addr) {
```
We're creating a new outbound future, but we immediately throw it away if it's not ready? There's no way this can work for something less trivial than mplex.

tomaka commented on core/src/connection_reuse.rs:
```diff
@@ -40,205 +40,556 @@
+                ..
+            } => {
+                if closed.load(Ordering::Acquire) {
+                    (PeerState::Errored(IoError::new(IoErrorKind::BrokenPipe, "Connection closed")),
```
If we open a connection, then later close it, then later attempt to open again, we will forever trigger this condition, no?
If I'm not mistaken, we should attempt to dial again instead.

tomaka commented on core/src/connection_reuse.rs:
```diff
@@ -40,205 +40,556 @@
+            };
+            self.new_con_notify();
+            // Althought we've just started in pending state, there are transports, which are
+            // immediately ready - namely 'memory' - so we need to poll our future right away
```
You always need to poll immediately, otherwise the current task will not be waken up later when it needs to.

tomaka commented on core/src/connection_reuse.rs:
```diff
@@ -40,205 +40,556 @@
+        }
+
+        for to_remove in to_remove {
+            conn.remove(&to_remove).map(|s| s.reset_conn());
```
I would really really really prefer `if let Some` over `map`

gnunicorn commented:
closing in favor of #446.
